app/models/stage.rb

```diff
@@ -28,6 +28,8 @@ class Stage < ActiveRecord::Base
   attr_writer :command
   before_save :build_new_project_command
 
+  before_soft_delete :verify_not_part_of_pipeline
```

Comment:
I would vote for copying the ```has_soft_deletion default_scope: true``` into the pipeline stage concern, rather then moving it to address this. We're trying really hard not to put plugin specific logic into the core code.